lib/Dialect/AIEVec/Transforms/ConvertVectorToAIEVec.cpp

```diff
@@ -1119,6 +1119,121 @@ using LowerVectorMinFOpToAIEVecMinOp =
 using LowerVectorMaxFOpToAIEVecMaxOp =
     LowerVectorMinMaxOpToAIEVecMinMaxOp<arith::MaxFOp, aievec::MaxOp>;
 
+static arith::CmpIPredicate
+convertToIntegerPredicate(arith::CmpFPredicate pred) {
+  switch (pred) {
+  case CmpFPredicate::UEQ:
+  case CmpFPredicate::OEQ:
+    return CmpIPredicate::eq;
+  case CmpFPredicate::UGT:
+  case CmpFPredicate::OGT:
+    return CmpIPredicate::ugt;
+  case CmpFPredicate::UGE:
+  case CmpFPredicate::OGE:
+    return CmpIPredicate::uge;
+  case CmpFPredicate::ULT:
+  case CmpFPredicate::OLT:
+    return CmpIPredicate::ult;
+  case CmpFPredicate::ULE:
+  case CmpFPredicate::OLE:
+    return CmpIPredicate::ule;
+  case CmpFPredicate::UNE:
+  case CmpFPredicate::ONE:
+    return CmpIPredicate::ne;
+  default:
+    llvm_unreachable("Unexpected predicate!");
+  }
+}
+
+static arith::CmpIPredicate
+convertToIntegerPredicate(arith::CmpIPredicate pred) {
+  return pred;
+}
+
+static aievec::CmpOp createCmpOpAieML(ConversionPatternRewriter &rewriter,
+                                      CmpIPredicate pred, Location loc,
+                                      Type type, Value lhs, Value rhs) {
+  switch (pred) {
+  case CmpIPredicate::eq:
+    return rewriter.create<aievec::CmpOp>(loc, type, lhs, rhs, "eq");
+  case CmpIPredicate::ne:
+    return rewriter.create<aievec::CmpOp>(loc, type, lhs, rhs, "ne");
+  case CmpIPredicate::slt:
+  case CmpIPredicate::ult:
+    return rewriter.create<aievec::CmpOp>(loc, type, lhs, rhs, "lt");
```

Comment:
This does not preserve semantics. slt and ult cannot both be mapped to "lt".
For arith.cmpi the operands are SignLess and the interpretation of whether they both are signed or whether they both are unsigned is specified by the predicate.

Reply:
The signless type in MLIR allows values to be treated as either signed or unsigned depending on the context. If the operands are signed, it is not allowed to use unsigned operator. The reverse is also true. In our aie-ml intrinsics, they are use the same operator for unsigned and signed comparison depends on the operands' data types. I believe the current setting is good for the semantics.

Reply:
@linay-xsj : in MLIR, i32 is a 32-bit integer type with no signedness semantics. See BuiltinTypes.td; operations like comparison require signedness semantics. For arith.cmp the signedness semantics are NOT provided by the operand type, but instead they are provided by the predicate (slt vs ult). When converting an arith.cmp to an aievec.comp, we are not changing the types of the operands. Consequently the signedness semantics must be conveyed via the predicate of the aievec.cmp. Thus, "signed less than" and "unsigned less than" must use different predicates.
@jsetoain : Could you also please comment on this.

Reply:
I agree with you, David. Coalescing `<s|u><cmp>` down to `<cmp>` loses signedness semantics. It just so happens to work here because we're lowering the comparison down to C++ from here, and the types will be "signful" in C++-land, which will allow the right function call to be invoked. But that's something that we break here, and gets coincidentally fixed in ***our*** _current_ next step. It may (and probably will) break in pretty much every other instance; e.g.: when lowering to LLVM.



> The signless type in MLIR allows values to be treated as either signed or unsigned depending on the context.

That is exactly right, and the context in this case is provided by the predicate. Removing signedness from the predicate breaks semantics. Information about signedness (whether the comparison is signed or unsigned) is lost, and we still want to keep that information at this level.

I understand we want to keep the translation simple, so we avoid "one to many" translations. But "many to one translations" do not add complexity to the translation.